File: app/src/main/java/com/nexus/browser/media/VideoHarvesterEngine.kt
package com.nexus.browser.media

import java.net.HttpURLConnection
import java.net.URI
import java.net.URL
import java.util.Locale
import java.util.regex.Pattern

/** Discovers publicly exposed video/player URLs. No DRM/login/CAPTCHA/access-control bypass. */
data class HarvestedVideo(
    val title: String,
    val url: String,
    val host: String,
    val type: String,
    val status: String,
    val selected: Boolean = true
)

object VideoHarvesterEngine {
    private const val MAX_HTML_BYTES = 5 * 1024 * 1024
    private const val MAX_LINKED_PAGES = 60
    private const val MAX_RESULTS = 1000

    private val attrPattern = Pattern.compile("(?:href|src|data-src|data-url|data-video|data-file|content)\\s*=\\s*['\\\"]([^'\\\"]+)['\\\"]", Pattern.CASE_INSENSITIVE)
    private val urlPattern = Pattern.compile("https?://[^\\s\\\"'<>\\\\]+", Pattern.CASE_INSENSITIVE)
    private val escapedUrlPattern = Pattern.compile("https?:\\\\?/\\\\?/[^\\s\\\"'<>]+", Pattern.CASE_INSENSITIVE)
    private val titlePattern = Pattern.compile("<title[^>]*>(.*?)</title>", Pattern.CASE_INSENSITIVE or Pattern.DOTALL)

    fun harvest(pageUrl: String, deepInspect: Boolean = true): List<HarvestedVideo> {
        val queue = ArrayDeque<Pair<String, String>>()
        val visited = HashSet<String>()
        val result = LinkedHashMap<String, HarvestedVideo>()
        queue.add(pageUrl to "Page")

        while (queue.isNotEmpty() && visited.size < MAX_LINKED_PAGES && result.size < MAX_RESULTS) {
            val (url, inheritedTitle) = queue.removeFirst()
            val normalizedPage = normalize(url) ?: continue
            if (!visited.add(normalizedPage)) continue
            val html = fetchHtml(normalizedPage) ?: continue
            val pageTitle = extractTitle(html).ifBlank { inheritedTitle }
            extractCandidates(html, normalizedPage).forEach { candidate ->
                val normalized = normalize(candidate) ?: return@forEach
                val type = classify(normalized)
                if (type == "LINK") return@forEach
                val host = runCatching { URI(normalized).host.orEmpty() }.getOrDefault("")
                val status = if (type in setOf("MP4", "WEBM", "M3U8", "MEDIA")) "MEDIA SOURCE" else "PLAYER / VIDEO PAGE"
                result.putIfAbsent(normalized, HarvestedVideo(pageTitle, normalized, host, type, status))
                if (deepInspect && type == "PLAYER" && queue.size + visited.size < MAX_LINKED_PAGES) queue.add(normalized to pageTitle)
            }
        }
        return result.values.toList()
    }

    private fun fetchHtml(url: String): String? = runCatching {
        val c = URL(url).openConnection() as HttpURLConnection
        c.connectTimeout = 12_000; c.readTimeout = 20_000; c.instanceFollowRedirects = true
        c.setRequestProperty("User-Agent", UA)
        c.setRequestProperty("Accept", "text/html,application/xhtml+xml,application/json,text/plain,*/*;q=0.8")
        try { c.inputStream.use { it.readNBytes(MAX_HTML_BYTES).toString(Charsets.UTF_8) } } finally { c.disconnect() }
    }.getOrNull()

    private fun extractCandidates(html: String, baseUrl: String): List<String> {
        val out = LinkedHashSet<String>()
        fun add(raw: String?) {
            if (raw.isNullOrBlank()) return
            val value = decode(raw.trim())
            if (value.startsWith("data:") || value.startsWith("javascript:") || value.startsWith("mailto:")) return
            runCatching { URI(baseUrl).resolve(value).toString() }.getOrNull()?.let(out::add)
        }
        val attrs = attrPattern.matcher(html)
        while (attrs.find()) add(attrs.group(1))
        val urls = urlPattern.matcher(html)
        while (urls.find()) add(urls.group())
        val escaped = escapedUrlPattern.matcher(html)
        while (escaped.find()) add(escaped.group().replace("\\/", "/"))
        return out.toList()
    }

    private fun normalize(url: String): String? = runCatching {
        val u = URI(url)
        if (u.scheme?.lowercase(Locale.US) !in setOf("http", "https") || u.host.isNullOrBlank()) null else u.toString()
    }.getOrNull()

    private fun classify(url: String): String {
        val p = runCatching { URI(url).path.orEmpty().lowercase(Locale.US) }.getOrDefault("")
        return when {
            p.endsWith(".mp4") || p.endsWith(".m4v") || p.endsWith(".mov") -> "MP4"
            p.endsWith(".webm") -> "WEBM"
            p.endsWith(".m3u8") -> "M3U8"
            p.endsWith(".ts") -> "MEDIA"
            isLikelyVideoPage(url) -> "PLAYER"
            else -> "LINK"
        }
    }

    private fun isLikelyVideoPage(url: String): Boolean {
        val s = url.lowercase(Locale.US)
        return listOf("/video", "/watch", "/embed", "/player", "/stream", "/trailer", "/episode", "/play", "videoplayer").any(s::contains)
    }

    private fun extractTitle(html: String): String = titlePattern.matcher(html).let {
        if (!it.find()) "" else decode(it.group(1)).replace(Regex("<[^>]+>"), "").replace(Regex("\\s+"), " ").trim().take(160)
    }

    private fun decode(value: String) = value.replace("&amp;", "&").replace("&quot;", "\"").replace("&#39;", "'").replace("\\/", "/")
    private const val UA = "Mozilla/5.0 (Android; NEXUS Browser/1.0)"
}

File: lib/engines/video_harvester_engine.dart
import 'dart:collection';
import 'dart:convert';
import 'dart:io';

import '../models/harvested_video.dart';
import 'manifest_parser.dart';
import 'network_sniffer.dart';

/// Discovers publicly exposed video/player URLs. No DRM/login/CAPTCHA/access-
/// control bypass. 1:1-Verhalten zu Kotlins `VideoHarvesterEngine`.
class VideoHarvesterEngine {
  VideoHarvesterEngine._();

  static const int _maxHtmlBytes = 5 * 1024 * 1024;
  static const int _maxLinkedPages = 60;
  static const int _maxResults = 1000;
  static const String _userAgent = 'Mozilla/5.0 (Linux; NEXUS Browser/1.0)';

  static final RegExp _attrPattern = RegExp(
    r'''(?:href|src|data-src|data-url|data-video|data-file|content)\s*=\s*['"]([^'"]+)['"]''',
    caseSensitive: false,
  );
  // Viele Player betten ihre Quelle als JSON-Konfiguration in einem
  // <script>-Block ein statt als HTML-Attribut, z.B. `"file": "…mp4"` oder
  // `"hls": "…m3u8"` in einem JWPlayer-/Video.js-Setup-Aufruf. Das obige
  // Attribut-Muster sieht so etwas nicht, weil es HTML-Attribut-Syntax
  // erwartet, keine JSON-Syntax.
  static final RegExp _jsonKeyPattern = RegExp(
    r'''"(?:file|src|source|url|hls|dash|contentUrl|videoUrl|streamUrl)"\s*:\s*"([^"]+)"''',
    caseSensitive: false,
  );
  static final RegExp _urlPattern = RegExp(
    r'''https?://[^\s"'<>\\]+''',
    caseSensitive: false,
  );
  static final RegExp _escapedUrlPattern = RegExp(
    r'''https?:\\?/\\?/[^\s"'<>]+''',
    caseSensitive: false,
  );
  static final RegExp _titlePattern = RegExp(
    r'<title[^>]*>(.*?)</title>',
    caseSensitive: false,
    dotAll: true,
  );

  static Future<List<HarvestedVideo>> harvest(
    String pageUrl, {
    bool deepInspect = true,
  }) async {
    final queue = Queue<_QueueItem>()..add(_QueueItem(pageUrl, 'Page'));
    final visited = <String>{};
    final result = <String, HarvestedVideo>{};

    while (queue.isNotEmpty &&
        visited.length < _maxLinkedPages &&
        result.length < _maxResults) {
      final item = queue.removeFirst();
      final normalizedPage = _normalize(item.url);
      if (normalizedPage == null || !visited.add(normalizedPage)) continue;

      final html = await _fetchHtml(normalizedPage);
      if (html == null) continue;

      final pageTitle = _extractTitle(html).isNotEmpty
          ? _extractTitle(html)
          : item.inheritedTitle;

      for (final found in _classifyHtml(html, normalizedPage, pageTitle)) {
        result.putIfAbsent(found.url, () => found);
        if (deepInspect &&
            found.type == 'PLAYER' &&
            queue.length + visited.length < _maxLinkedPages) {
          queue.add(_QueueItem(found.url, pageTitle));
        }
      }
    }
    return result.values.toList();
  }

  /// Wertet bereits vorliegendes HTML aus, statt es selbst per HTTP zu
  /// holen — keine Rekursion in verlinkte Seiten. Der entscheidende
  /// Anwendungsfall: [html] kann das **nach JavaScript-Ausführung
  /// gerenderte** DOM einer echten WebView sein (`document.documentElement.
  /// outerHTML`), nicht nur die rohe Server-Antwort. Viele Streaming-Seiten
  /// bauen ihren `<video>`/Player-Quelltext erst per JS zusammen, nachdem
  /// die Seite geladen ist — ein reiner HTTP-Fetch (wie [harvest] ihn für
  /// verlinkte Seiten macht) sieht diese URLs nie, weil sie in der
  /// Server-Antwort schlicht noch nicht existieren. Das war bei genauerer
  /// Betrachtung vermutlich die eigentliche Ursache dafür, dass zuletzt kaum
  /// echte Videodateien gefunden wurden, nicht nur ein zu schwacher
  /// Downloader.
  static List<HarvestedVideo> extractFromRenderedHtml(
    String html,
    String baseUrl, {
    String titleHint = '',
  }) {
    final title =
        _extractTitle(html).isNotEmpty ? _extractTitle(html) : titleHint;
    return _classifyHtml(html, baseUrl, title);
  }

  static List<HarvestedVideo> classifyUrls(
    List<String> urls,
    String pageTitle, {
    String pageUrl = '',
    String referrer = '',
    String userAgent = '',
  }) {
    return classifyCaptures(urls.map((url) => MediaCapture(
          url: url, pageUrl: pageUrl, referrer: referrer, userAgent: userAgent,
        )).toList(), pageTitle);
  }

  static List<HarvestedVideo> classifyCaptures(
    List<MediaCapture> captures,
    String pageTitle,
  ) {
    final out = <HarvestedVideo>[];
    final seen = <String>{};
    for (final capture in captures) {
      final raw = capture.url;
      if (!seen.add(raw)) continue;
      final normalized = _normalize(raw);
      if (normalized == null) continue;
      final type = _classify(normalized);
      if (type == 'LINK' || type == 'PLAYER') continue;
      String host;
      try {
        host = Uri.parse(normalized).host;
      } catch (_) {
        host = '';
      }
      out.add(HarvestedVideo(
        title: pageTitle,
        url: normalized,
        host: host,
        type: type,
        status: type == 'BLOB'
            ? 'MEDIA SOURCE · ${capture.source} (blob: — nicht direkt herunterladbar)'
            : 'MEDIA SOURCE · ${capture.source}',
        source: capture.source,
        pageUrl: capture.pageUrl,
        referrer: capture.referrer,
        userAgent: capture.userAgent,
        headers: capture.headers,
        cookies: capture.cookies,
      ));
    }
    return out;
  }

  static List<HarvestedVideo> _classifyHtml(
    String html,
    String baseUrl,
    String pageTitle,
  ) {
    final out = <HarvestedVideo>[];
    for (final candidate in _extractCandidates(html, baseUrl)) {
      final normalized = _normalize(candidate);
      if (normalized == null) continue;
      final type = _classify(normalized);
      if (type == 'LINK') continue;
      String host;
      try {
        host = Uri.parse(normalized).host;
      } catch (_) {
        host = '';
      }
      final status = type == 'BLOB'
          ? 'MEDIA SOURCE (blob: — nicht direkt herunterladbar)'
          : {'MP4', 'WEBM', 'M3U8', 'MEDIA', 'DASH'}.contains(type)
              ? 'MEDIA SOURCE'
              : 'PLAYER / VIDEO PAGE';
      out.add(HarvestedVideo(
        title: pageTitle,
        url: normalized,
        host: host,
        type: type,
        status: status,
        source: 'DOM',
        pageUrl: baseUrl,
      ));
    }
    return out;
  }

  /// Parses public HLS/DASH manifests and attaches quality variants.
  static Future<HarvestedVideo> enrichManifest(HarvestedVideo video) async {
    if (video.type != 'M3U8' && video.type != 'DASH') return video;
    final variants = await ManifestParser.inspect(video.url, headers: {
      ...video.headers,
      if (video.referrer.isNotEmpty) 'Referer': video.referrer,
      if (video.userAgent.isNotEmpty) 'User-Agent': video.userAgent,
      if (video.cookies.isNotEmpty) 'Cookie': video.cookies,
    });
    if (variants.isEmpty) return video;
    final best = variants.reduce((a, b) => a.bandwidth >= b.bandwidth ? a : b);
    return video.copyWith(
      variants: variants,
      quality: best.quality,
      status: '${video.status} · ${variants.length} Qualitätsstufen',
    );
  }

  static Future<String?> fetchHtmlForDebug(String url) => _fetchHtml(url);

  static Future<String?> _fetchHtml(String url) async {
    final client = HttpClient()
      ..connectionTimeout = const Duration(seconds: 12)
      ..autoUncompress = true;
    try {
      final request = await client
          .getUrl(Uri.parse(url))
          .timeout(const Duration(seconds: 12));
      request.headers.set(HttpHeaders.userAgentHeader, _userAgent);
      request.headers.set(
        HttpHeaders.acceptHeader,
        'text/html,application/xhtml+xml,application/json,text/plain,*/*;q=0.8',
      );
      final response =
          await request.close().timeout(const Duration(seconds: 20));
      final bytes = <int>[];
      await for (final chunk in response) {
        bytes.addAll(chunk);
        if (bytes.length >= _maxHtmlBytes) break;
      }
      return utf8.decode(bytes, allowMalformed: true);
    } catch (_) {
      return null;
    } finally {
      client.close(force: true);
    }
  }

  static List<String> _extractCandidates(String html, String baseUrl) {
    final out = <String>{};
    void add(String? raw) {
      if (raw == null || raw.trim().isEmpty) return;
      final value = _decode(raw.trim());
      if (value.startsWith('data:') ||
          value.startsWith('javascript:') ||
          value.startsWith('mailto:')) {
        return;
      }
      try {
        out.add(Uri.parse(baseUrl).resolve(value).toString());
      } catch (_) {
        // ignorieren, wie im Kotlin-Original (runCatching { }.getOrNull())
      }
    }

    for (final m in _attrPattern.allMatches(html)) {
      add(m.group(1));
    }
    for (final m in _jsonKeyPattern.allMatches(html)) {
      add(m.group(1)?.replaceAll(r'\/', '/'));
    }
    for (final m in _urlPattern.allMatches(html)) {
      add(m.group(0));
    }
    for (final m in _escapedUrlPattern.allMatches(html)) {
      add(m.group(0)?.replaceAll(r'\/', '/'));
    }
    return out.toList();
  }

  static String? _normalize(String url) {
    try {
      final u = Uri.parse(url);
      final scheme = u.scheme.toLowerCase();
      if (scheme == 'blob') {
        // blob:https://origin/uuid — kein normaler Host-Teil, aber ein
        // starkes, eigenständiges Signal (siehe _classify: wird als
        // eigener, ausdrücklich NICHT herunterladbarer Typ markiert statt
        // stillschweigend verworfen zu werden).
        return url;
      }
      if ((scheme != 'http' && scheme != 'https') || u.host.isEmpty) {
        return null;
      }
      return u.toString();
    } catch (_) {
      return null;
    }
  }

  static String _classify(String url) {
    if (url.toLowerCase().startsWith('blob:')) return 'BLOB';
    String path;
    try {
      path = Uri.parse(url).path.toLowerCase();
    } catch (_) {
      path = '';
    }
    const directExtensions = {
      '.mp4', '.m4v', '.mov', '.webm', '.mkv', '.flv', '.3gp', '.ogv', //
    };
    if (path.endsWith('.mp4') || path.endsWith('.m4v') || path.endsWith('.mov')) {
      return 'MP4';
    }
    if (path.endsWith('.webm')) return 'WEBM';
    if (directExtensions.any(path.endsWith)) return 'MEDIA';
    if (path.endsWith('.m3u8')) return 'M3U8';
    if (path.endsWith('.mpd')) return 'DASH';
    if (path.endsWith('.ts')) return 'MEDIA';
    if (_isLikelyVideoPage(url)) return 'PLAYER';
    return 'LINK';
  }

  static bool _isLikelyVideoPage(String url) {
    final s = url.toLowerCase();
    const markers = [
      '/video', '/watch', '/embed', '/player', '/stream', '/trailer',
      '/episode', '/play', 'videoplayer', //
    ];
    return markers.any(s.contains);
  }

  static String _extractTitle(String html) {
    final match = _titlePattern.firstMatch(html);
    if (match == null) return '';
    final decoded = _decode(match.group(1) ?? '')
        .replaceAll(RegExp(r'<[^>]+>'), '')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim();
    return decoded.length > 160 ? decoded.substring(0, 160) : decoded;
  }

  static String _decode(String value) => value
      .replaceAll('&amp;', '&')
      .replaceAll('&quot;', '"')
      .replaceAll('&#39;', "'")
      .replaceAll(r'\/', '/');
}

class _QueueItem {
  final String url;
  final String inheritedTitle;
  const _QueueItem(this.url, this.inheritedTitle);
}
